docs/developer/architecture/ci.rst

```diff
@@ -0,0 +1,120 @@
+CI
+==
+
+Pre merge
+---------
+
+What should be tested in pre-merge on every ``user/*``, ``feature/*``,
+``improvement/*``, ``bugfix/*``, ``w/*`` branches.
+
+- Building tests:
+
+  - Build
+  - Lint
+  - Unit tests
+
+- Installation tests
+
+  - Simple install RHEL
+  - Simple install CentOs + expansion
+
+When merging several PRs at the same time, so when we are on a ``q/*``
+branches, we may want to have additionnal test like:
+
+- Simple upgrade/downgrade
+
+Pre-merge shouldn't be too long as it run on every push on the feature
+branches.
+
+.. _post-merge:
+
+Post merge
+----------
+
+On each and every ``development/2.*`` branches we want to run complex tests to
+ensure that the current developmnent product is well working.
+
+Nightly
+^^^^^^^
+
+- Upgrade, downgrade tests:
+
+  - For development branch just before
+
+    e.g.: on ``development/2.x`` test upgrade from ``development/2.(x-1)``
+    and downgrade to ``development/2.(x-1)``
+
+    - Build branch ``development/2.(x-1)``
+    - Tests:
+
+      - Single node test
+      - Complex deployment test
+
+  - For last released version of current minor
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.x.(y-1)`` and downgrade to ``metalk8s-2.x.(y-1)``
+
+    - Single node test
+    - Complex deployment test
+
+  - For last released version of minor just before
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.(x-1).z`` and downgrade to ``metalk8s-2.(x-1).z`` where
+    "2.(x-1).z" is the last patch released version for "2.(x-1)" (``z`` may be
+    different from ``y``)
+
+    - Single node test
+    - Complex deployment test
+
+- Backup, restore tests
+- Solutions tests
+
+.. todo::
+
+   - Define what is "complex deployment"
+
+Weekly
+^^^^^^
+
+More complex tests (TBD)
+
+Adaptive test plan
+------------------
+
+CI pre-merge may be more flexible by including some logic about the content
+of the changeset.
+
+For example :
+
+- for a commit changing only documentation we don't need/want to run
+  all the test about installation but just documentation part.
+- for a commit touching upgrade orchestrate we want to test upgrade directly
+  in pre-merge and not wait :ref:`Post merge<post-merge>` build to get the
```

Comment:
- Please begin the sentences in capitals (For)
- We could rewrite the first sentence to read
`For a commit that changes solely/uniquely documentation, we don't need/want to run all the installation test but rather run tests related to documentation`